### PR TITLE
Python Service Reflection

### PR DESCRIPTION
**Pull request by mehrdada**
Implements the standard reflection service for gRPC python

### CODE REVIEW

**Comment by mehrdada** [comment]
Tests on Windows and Mac fail seemingly as a result of https://github.com/google/protobuf/issues/2882

File: src/python/grpcio_reflection/grpc_reflection/v1alpha/reflection.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Copyright 2016, Google Inc.
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_reflection/grpc_reflection/v1alpha/reflection.py]: Project policy is that we don't update copyright year when modifying existing files (but new files, including branches of old files, get the current year).

I apologize that this isn't written down anywhere, I don't think. :-(

File: src/python/grpcio_reflection/grpc_reflection/v1alpha/reflection.py [hunk 2]
```diff
@@ -84,13 +84,18 @@ def _file_containing_symbol(self, fully_qualified_name):
...
-    def _file_containing_extension(containing_type, extension_number):
-        # TODO(atash) Python protobuf currently doesn't support querying extensions.
-        # https://github.com/google/protobuf/issues/2248
-        return reflection_pb2.ServerReflectionResponse(
-            error_response=reflection_pb2.ErrorResponse(
-                error_code=grpc.StatusCode.UNIMPLEMENTED.value[0],
-                error_message=grpc.StatusCode.UNIMPLMENTED.value[1].encode(),))
+    def _file_containing_extension(self, containing_type, extension_number):
+        try:
+            pool = self._pool
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_reflection/grpc_reflection/v1alpha/reflection.py]: Why this local alias? Why not just use `self._pool` throughout this behavior?

File: src/python/grpcio_tests/tests/reflection/_reflection_servicer_test.py [hunk 3]
```diff
@@ -50,6 +50,7 @@
 _SERVICE_NAMES = ('Angstrom', 'Bohr', 'Curie', 'Dyson', 'Einstein', 'Feynman',
                   'Galilei')
 _EMPTY_EXTENSIONS_SYMBOL_NAME = 'grpc.testing.proto2.EmptyWithExtensions'
+_EMPTY_EXTENSIONS_NUMBERS = (124, 125, 126, 127, 128)
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_tests/tests/reflection/_reflection_servicer_test.py]: Include the ostensibly-optional-but-always-a-good-idea last-position comma after `128`.

File: src/python/grpcio_reflection/grpc_reflection/v1alpha/reflection.py [hunk 4]
```diff
@@ -142,3 +143,14 @@ def ServerReflectionInfo(self, request_iterator, context):
...
                         error_message=grpc.StatusCode.INVALID_ARGUMENT.value[1]
                         .encode(),))
+
+
+def enable_server_reflection(service_names, server):
+    """Enables server reflection on a server.
+
+    Args:
+      service_names: Iterable of fully-qualified service names available.
+      server: Server to which reflection service will be added.
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_reflection/grpc_reflection/v1alpha/reflection.py]: "`grpc.Server` to which..."

File: src/python/grpcio_tests/tests/reflection/_reflection_servicer_test.py [hunk 5]
```diff
@@ -39,14 +39,18 @@
 from google.protobuf import descriptor_pool
 from google.protobuf import descriptor_pb2
 
-from src.proto.grpc.testing.proto2 import empty2_extensions_pb2
 from src.proto.grpc.testing import empty_pb2
+
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_tests/tests/reflection/_reflection_servicer_test.py]: Why this blank line?

File: src/python/grpcio_tests/tests/reflection/_reflection_servicer_test.py [hunk 6]
```diff
@@ -39,14 +39,18 @@
...
+from src.proto.grpc.testing.proto2 import empty2_pb2
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_tests/tests/reflection/_reflection_servicer_test.py]: Unused import?

File: src/python/grpcio_reflection/grpc_reflection/v1alpha/reflection.py [hunk 7]
```diff
@@ -84,13 +84,17 @@ def _file_containing_symbol(self, fully_qualified_name):
...
             return _file_descriptor_response(descriptor)
 
-    def _file_containing_extension(containing_type, extension_number):
-        # TODO(atash) Python protobuf currently doesn't support querying extensions.
-        # https://github.com/google/protobuf/issues/2248
-        return reflection_pb2.ServerReflectionResponse(
-            error_response=reflection_pb2.ErrorResponse(
-                error_code=grpc.StatusCode.UNIMPLEMENTED.value[0],
-                error_message=grpc.StatusCode.UNIMPLMENTED.value[1].encode(),))
+    def _file_containing_extension(self, containing_type, extension_number):
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_reflection/grpc_reflection/v1alpha/reflection.py]: Adding one function is rather small for a single commit; you could fold this into the next commit if you wanted to.

File: src/python/grpcio_tests/tests/reflection/_reflection_servicer_test.py (hunk 6 above)
Comment (mehrdada) [inline src/python/grpcio_tests/tests/reflection/_reflection_servicer_test.py]: It is unused directly in the file, but it has side effects.

Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_tests/tests/reflection/_reflection_servicer_test.py]:
```py
# empty2_pb2 is imported for import-consequent side-effects.
from src.proto.grpc.testing.proto2 import empty2_pb2  # pylint: disable=unused-import
```
in that case, please?

File: src/python/grpcio_reflection/grpc_reflection/v1alpha/reflection.py (hunk 7 above)
Comment (mehrdada) [inline src/python/grpcio_reflection/grpc_reflection/v1alpha/reflection.py]: I am a huge advocate for small commits. There are a number of advantages to small commits and almost none to bigger ones. _I could go on but the margin on this page is too small for that..._ :)

File: src/python/grpcio_reflection/grpc_reflection/v1alpha/reflection.py (hunk 1 above)
Comment (mehrdada) [inline src/python/grpcio_reflection/grpc_reflection/v1alpha/reflection.py]: done

File: src/python/grpcio_reflection/grpc_reflection/v1alpha/reflection.py (hunk 2 above)
Comment (mehrdada) [inline src/python/grpcio_reflection/grpc_reflection/v1alpha/reflection.py]: I preferred not repeating `self._pool` on subsequent lines, but don't feel too strongly about it. done.

File: src/python/grpcio_tests/tests/reflection/_reflection_servicer_test.py (hunk 3 above)
Comment (mehrdada) [inline src/python/grpcio_tests/tests/reflection/_reflection_servicer_test.py]: done

File: src/python/grpcio_tests/tests/reflection/_reflection_servicer_test.py (hunk 5 above)
Comment (mehrdada) [inline src/python/grpcio_tests/tests/reflection/_reflection_servicer_test.py]: done

File: src/python/grpcio_tests/tests/reflection/_reflection_servicer_test.py [hunk 8]
```diff
@@ -40,8 +40,8 @@
 from google.protobuf import descriptor_pb2
 
 from src.proto.grpc.testing import empty_pb2
-
-from src.proto.grpc.testing.proto2 import empty2_pb2
+# empty2_pb2 is imported for import-consequent side-effects.
+from src.proto.grpc.testing.proto2 import empty2_pb2 # pylint: disable=unused-import
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_tests/tests/reflection/_reflection_servicer_test.py]: Typographical minute nit-pick: [two spaces between code and beginning-of-comment octothorpe](https://google.github.io/styleguide/pyguide.html?showone=Whitespace#Whitespace).

Comment (mehrdada) [inline src/python/grpcio_tests/tests/reflection/_reflection_servicer_test.py]: I copied and pasted directly from your https://github.com/grpc/grpc/pull/10216#discussion_r107531453. done 👍

File: src/python/grpcio_tests/tests/reflection/_reflection_servicer_test.py (hunk 6 above)
Comment (mehrdada) [inline src/python/grpcio_tests/tests/reflection/_reflection_servicer_test.py]: done

**Comment by nathanielmanistaatgoogle** [comment]
Looks good content-wise.

Any desire to amend your earlier commits to meet [the commit message guidelines](https://chris.beams.io/posts/git-commit/#seven-rules)?

Will approve following test results.

**Comment by mehrdada** [comment]
I looked at the project history and it seems we do not follow any guidelines strictly for our commit messages, so I decided to leave it as is this time as it necessitates another `push -f` which will ruin our inline comment context on this pull request. I did, however, modify the title of the last commit at the time (which could have been amended) as that was particularly obnoxious (long title).

**Comment by nathanielmanistaatgoogle** [comment]
@mehrdada: yeah, the commit log message guidelines have thus far not yet spread beyond the Python sub-team. There are some ways in which I've pushed to make the Python sub-codebase an especially well-tended garden, and this is one of them. 😄

**Comment by mehrdada** [comment]
Oh well... `yapf` is now complaining about a single character, and not any character for that matter, a single space, and it is strictly a worse suggestion. Committing a patch...

**Comment by mehrdada** [comment]
Test failures:

- Bazel Full Build & Tests: not yet in effect
- Sanity Tests (internal CI, ignore for now): ignoring for now

**Pull request merged**